net/net.go

```diff
@@ -70,15 +70,23 @@ func AddIPOffset(base *big.Int, offset int) net.IP {
 // RangeSize returns the size of a range in valid addresses.
 // returns the size of the subnet (or math.MaxInt64 if the range size would overflow int64)
 func RangeSize(subnet *net.IPNet) int64 {
-	ones, bits := subnet.Mask.Size()
-	if bits == 32 && (bits-ones) >= 31 || bits == 128 && (bits-ones) >= 127 {
-		return 0
-	}
+	size := RangeSizeU64(subnet)
 	// this checks that we are not overflowing an int64
-	if bits-ones >= 63 {
+	if size >= math.MaxInt64 {
 		return math.MaxInt64
 	}
-	return int64(1) << uint(bits-ones)
+	return int64(size)
+}
+
+// RangeSizeU64 returns the size of a range in valid addresses.
+// returns the size of the subnet (or math.MaxInt64 if the range size would overflow int64)
+func RangeSizeU64(subnet *net.IPNet) uint64 {
```

Comment:
There's not much argument for larger than /64 pod subnets, but if we think these functions might be used by the code that allocates CIDRs to nodes (rather than the code that allocates IPs to pods) then it might need to deal with, eg, allocating /64s out of a /56.

(Of course, such code wouldn't actually have to be working with more than 64 bits *at a time*; it would just, eg, be allocating 8-bit numbers that get plugged into bits 64-71 of an IP address template, rather than allocating 64-bit numbers that get plugged into bits 0-63 of an IP address template.)

Reply:
So what do we want to do with this?  Document that we only support up to /64, or go bignum or uint128 ?

Reply:
if this is a library we should aim for getting it right, I vote `bignum or uint128`

This is pretty neat https://golang.google.cn/src/net/netip/uint128.go

Reply:
having Uint128 can be very useful for playing with ipv6 addresses, and will save a lot of tooling, let me see how far I go with the uint128 type

Reply:
Why is that .cn?  It looks like it is aimed at bitmasks.

We also use big.Int in pkg/registry/core/service/ipallocator/bitmap.go which you could wrap as Uint128 or even just leave - nice because it is in stdlib.

Reply:
.cn no idea, first hit looking for uint128 golang.

>  nice because it is in stdlib.

uint128 was added with the netip.Addr package for performance https://tailscale.com/blog/netaddr-new-ip-type-for-go/ , but we may not care much about that, do we?

Reply:
but they are not suporting it as a public API - it's an impl detail

Reply:
Go should really have a native uint128 type.  https://github.com/golang/go/issues/9455

Reply:
`big.Int` doesn't work well with bit ops because it has no inherent length. (If you create a `big.Int` from `net.ParseIP("::ffff:ffff:ffff:ffff")` and then call `.Not()` on it, you get back `0`, not `ffff:ffff:ffff:ffff::`, because it wasn't paying attention to the fact that you considered the number to be 128 bits.)

But you don't need 90% of `big.Int` for handling IP addresses anyway. (When was the last time you multiplied an IP address?) [I wrote some code once to do IP/CIDR math](https://github.com/danwinship/sdn/blob/networkpolicy-ranges-ipv6/pkg/util/ranges/fixedint.go). It's not that big. (This was part of a package for [converting a NetworkPolicy `ipBlock` with `except`s into an equivalent set of CIDRs with no `except`s](https://github.com/danwinship/sdn/blob/networkpolicy-ranges-ipv6/pkg/util/ranges/networkpolicy_test.go#L125), so I only implemented the math I needed for that...)